Added emphasis

Hyreon opened the pull request:
Added emphasis

Hyreon committed:
Added emphasis

Added emphasis

danieltjewett commented on readme.md:
```diff
@@ -85,4 +85,4 @@ If you file an issue, the issue should contain a title and a clear description o
 ## License
 
-Draw Container is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT).
+GMS Tasks is open-sourced software licensed under the [MIT license](http://opensource.org/licenses/MIT).
```
Somehow this got changed with the old label.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -538,4 +607,12 @@ function draw_container(data)
+	
+	if (halign == fa_right) return -right_width;
+	else if (halign == fa_center) return (left_width-right_width)/2;
+	else return left_width; // fa_left or undefined
```
It seems like this is only used once.  Though we agree with moving out functionality into smaller chunks, we need `draw_container` to be performant, and extra nested functions will be slightly less.  We also try to keep one global function per file.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +465,94 @@ function draw_container(data)
+			
 		}
+		
+		show_debug_message(data.decorators);
```
`log` could be used here.  However, let's remove this.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -1,7 +1,16 @@
+	
+	var last_decorators = {
+		color: c_orange	
+	}
```
GML automatically cleans up garbage.  However, we use `delete` after we are done with structs with the hope that someday, `delete` will delete memory immediately: https://github.com/YoYoGames/GameMaker-Bugs/issues/3272

danieltjewett replied:
`draw_container` also recursively calls itself, so this could a lot of constructing for a simple property.  If this is needed, maybe a simple var would be better.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -35,6 +44,13 @@ function draw_container(data)
+	if (!struct_exists(data, "len"))
+	{
+		data.len = string_length(data.str);
+	}
```
This _might_ be able to be moved to draw_container_get_properties

danieltjewett commented:
For full transparency, this additional work is a commissioned project, for the needs of our internal game.  There are some pretty cool / powerful additions to this, but might be too much for the needs / performance of this project.  Feel free to fork / branch if those features are what you need.  Review coming soon.

danieltjewett commented on objects/obj_Message5/Draw_64.gml:
```diff
@@ -33,6 +33,9 @@ if (menuSize > 0)
 			grid: "row",
+			decorators: [{
+				color: c_blue	
+			}]
```
This is a pretty powerful feature, but I think it is too much.  I think we could simplify the parsing / etc., which just one key called `emphasisColor`, instead of a list of colors.

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,100 @@
+function update_data_with_components(data) {
+	data.components = get_components(data.str);
+	data.plaintext = string_join_ext("", array_map(data.components, get_contents));
+}
```
I think this idea is great, but beyond the needs of this project.

Thinking about this more, I think a simple `*` can be used for emphasis.  For example, `I am going to mark the location of the *secret location* on your map.`

The parser could simply be (not tested, not pseudo code, probably needs more thought):
```
var strs = [];
var currentStr = "";

var emphasisFlag = false;

var strLength = string_length(originalStr);
for (var i=1; i<=strLength; i++)
{
    var char = string_char_at(originalStr, i);

    if (char == "*")
    {
        if (currentStr != "")
        {
            array_push(strs, {
                str: currentStr,
                emphasis: emphasisFlag,
            });
        }

        emphasisFlag = !emphasisFlag;

        currentStr = "";
    }
    else
    {
        currentStr += char;
    }
}

if (emphasisFlag)
{
    throw "Unclosed '*'";
}
```

This could be added in the `str` section in `draw_container_get_properties`.  Check to see if `str` is not `""` first, and then create the `strs`, which is equivalent to the `component_get_lines` function.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +465,94 @@ function draw_container(data)
+	
+	var total_lines = array_length(lines);
+	
+	for (var line_index = 0; line_index < array_length(lines); line_index++)
```
I need to double check this in my other projects, but based on the screenshots sent on discord, it seems like the math is all good here.  This was the harder part of the project, so nice work!

danieltjewett replied:
I will double check after the updates have been made from the this review.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +465,94 @@ function draw_container(data)
-			draw_text_ext(startX, startY - 1, data.str, data.lineHeight, data.renderWidth);
-			draw_text_ext(startX, startY + 1, data.str, data.lineHeight, data.renderWidth);
+			var text_to_draw = segment.contents;
+			//text_to_draw = string_replace_all(text_to_draw," ","_");
```
We could remove this.

Hyreon replied:
I've removed the comment. Did you also want to remove text_to_draw = segment.contents?

danieltjewett replied:
I'm pretty sure we need that.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +465,94 @@ function draw_container(data)
+				if (data.textAlpha != 0 && data.computedOpacity != 0)
+				{
+					draw_set_alpha(data.textAlpha * data.computedOpacity);
+					if (array_length(segment.tags) == 0) {
+						draw_set_color(data.textColor);
+					} else {
```
There are many places of inconsistent formatting.  This project uses whitespace liberally.  For example:
```
if (expression)
{
    someCode
}
else if (anotherExpression)
{
    someMoreCode
}
else
{
    evenMoreCode
}
```
One rule (whether spoken, using a linter, or unspoken) when working on projects is to conform to that projects formatting style.  It's a pain (I have to use a style for my day job that isn't what I like), but it is best for the project and its future readers to be consistent.

Hyreon replied:
Found and replaced all instances of a `) {` `else {` or } else` pair to remove Egyptian braces.
If there are other formatting corrections you'd like to see, or anything I missed, let me know.

danieltjewett commented on scripts/draw_container_calculate/draw_container_calculate.gml:
```diff
@@ -2,6 +2,9 @@
 function draw_container_calculate(data)
 {	
+	//ensure the data is complete
+	update_data_with_components(data);
```
I think with the simpler direction of using just `*`, we don't need this here now.

Hyreon replied:
It didn't need to be there in the first place, it's a duplicate. Good catch.
However, it is required to remove get plaintext without asterisks, otherwise the width and height will be wrong. Sure, we could use `string_replace_all`, but this should be more performant (it takes advantage of the string splitting we've already done)

danieltjewett replied:
I'm confused, it seems to be removed now?

Hyreon replied:
There were two called to update_data_with_components. This was a duplicate. Now, it is (correctly) only called once.

danieltjewett commented on scripts/draw_container_get_properties/draw_container_get_properties.gml:
```diff
@@ -134,6 +134,11 @@ function draw_container_get_properties(data)
 		data.paddingBottom = data.padding;
 	}
 	
+	if (!variable_struct_exists(data, "decorators"))
+	{
+		data.decorators = [];
+	}
+	
```
For simplicity, we should simply call this `emphasisColor`, and default to `c_black` (inverse of `textColor`)

Hyreon replied:
Done (with `c_orange` as the default)

danieltjewett commented on scripts/component_emphasize/component_emphasize.gml:
```diff
@@ -0,0 +1,25 @@
+// Script assets have changed for v2.3.0 see
+// https://help.yoyogames.com/hc/en-us/articles/360005277377 for more information
+function component_emphasize(str, substr = "", tag = -1) {
+	
+	if (substr == "") {
+		substr = str;
+	}
+	
+	var last_pos = 0;
+	while (true) {
+		var pos = string_pos_ext(substr,str,last_pos);
+		if (pos == 0) break;
+		last_pos = pos + string_length(substr) + 4;
+		str = string_insert("}}",str,pos + string_length(substr));
+		if (tag != -1) {
+			str = string_insert("|",str,pos);
+			str = string_insert(string(tag),str,pos);
+			last_pos += string_length(string(tag)) + 1;
+		}
+		str = string_insert("{{",str,pos);
+	}
+	
+	return str;
+	
+}
```
I think with the simpler direction, this may not be needed.

Hyreon replied:
It's never needed. This is an optional function that can be used to programmatically emphasize a word or phrase. If you never use it, the game won't run it and it won't slow anything down. If you're sure you will absolutely never need to programmatically emphasize something, the program can be removed without any negative consequences.

danieltjewett replied:
I think `draw_container` should be the only way we interface with this system.  Therefore, I think we can remove this.

Hyreon replied:
This doesn't actually interface with this system - all it does is put asterisks around the found strings. It's a preprocessing step.
We can still remove this!

danieltjewett replied:
Yes, let's remove this.

danieltjewett reviewed (changes requested):
The screenshots sent on Discord seem to show that mathematically, the positioning is working well, which is great!  I need to confirm this in the Game Maker project, as well as make sure this doesn't break existing `draw_container` calls in our internal game.  I haven't done this yet because I do have some requests (which may alter some of the data structures / math).

Overall, I think the features added here are really powerful, but beyond the needs for what I envision for `draw_container`.  I don't see needing nested decorators, or multiple decorators as well.  Therefore, I think there is a lot of complexity that can be stripped out to make this more performant as well.  Since this is a recursive function, the less we can do, the better in the long run.  In the review, I provided an alternative parser to keep things simpler / performant.

With that said, I know this is a commissioned project, for _our_ needs.  Feel free to fork this and do what you need for yours :)

Hyreon committed:
Removed multiple decorators

Also improved code consistency across the board

Hyreon committed:
Replaced {{tags}} with *emphasis*

Lots of things broke as a result of this change, not entirely sure why. Debugging time

Hyreon committed:
Fixed sizing bug

data.plaintext is now the correct value

Hyreon committed:
Removed duplicate update_data_with_components

Hyreon committed:
Removed outdated TODOs

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,56 @@
+/// @description get_components(str) A function that breaks down a string into components
+/// @param str The string (or substring)
+
+// Components are formed by using {{two curly braces around the component.}}
+// Single curly braces {} have no effect for components.
+// If you have to put two curly braces in the actual string, use a non-rendering character between them.
+// Components can be nested within each other. {{This sentence could be gold {{and}} have an italicized 'and'.}}
+// To handle {{2|foreign languages}} switching the order of emphasized text, you can use {{1|tags}} to hard-code the order.
+// If no tags are used, the next decorator is used, starting with #1.
+// This is supposed to be a very resilient system - even if the string is formatted incorrectly, it shouldn't cause a game crash.
+// Despite this, unclosed and overlapping curly braces are undefined behavior, and I strongly recommend against them.
+// - Hyreon
```
This comment is outdated now.

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,56 @@
+
+function get_components(str)
+{
+	//var components = [{contents: str, emphasis: false}];
+	//return components;
```
We can probably remove.

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,56 @@
+	var i = 0;
+	while (i < string_length(str))
+	{
+		var skip = false;
```
`skip` seems unused here.

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,56 @@
+		var skip = false;
+		var char = string_char_at(str,++i);
+		
+		if (char == "*")
+		{
+			array_push(components, {
+                contents: buffer,
+                emphasis: emphasisFlag,
+            });
```
Some formatting issues (probably Game Maker / copy / paste / Windows)

Hyreon replied:
Looks like tabs aren't showing up on GitHub as they are in the GameMaker editor. Which is good, we don't want tabs and spaces mixing. It's all spaces now

danieltjewett commented on scripts/get_components/get_components.gml:
```diff
@@ -0,0 +1,56 @@
+		}
+	}
+	
+	array_push(components, {
+        contents: buffer,
+        emphasis: emphasisFlag,
+    });
```
Another formatting issue.

danieltjewett commented on scripts/update_data_with_components/update_data_with_components.gml:
```diff
@@ -0,0 +1,5 @@
+function update_data_with_components(data)
+{
+	data.components = get_components(data.str);
+	data.plaintext = string_join_ext("", array_map(data.components, get_component_contents));
```
I think this could be
```
function(component) {
   return component.contents;
});
```
And then, we can remove the global function `get_component_contents`, as it seems to be only used once.

danieltjewett commented on objects/obj_Message7/Draw_64.gml:
```diff
@@ -3,9 +3,10 @@
-var text = currentMessage;
-var index = floor(clamp(characterCounter, 0, currentMessageLength));
-var printText = string_copy(text, 1, index);
+//you can manually send in the modified text, but this doesn't play nice with emphasized text.
+//var text = currentMessage;
+//var index = floor(clamp(characterCounter, 0, currentMessageLength));
+//var printText = string_copy(text, 1, index);
```
Can you help me understand what these changes were for?  I don't see any use of `characterCounter` in this file on `72` below.

Hyreon replied:
When using the `characterCounter` provided, you'll have brief blips where there are incomplete emphases. For instance:
`*Chipot` or, worse, `*Chipotle is some of the best food in the world. Did you know the more *`.
By passing `len`, and determining what the plaintext is, we can remove any and all cases of junk input by:
1. beginning to draw the full string
2. once enough characters are drawn, stop drawing.

I'll remove the unused code.

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,130 @@
+// Breaks the text to be displayed into lines consisting of segments; used to display items correctly when middle-aligned or right-aligned
```
I completely missed this file last time.  This is where all the magic is happening!  After this second review, I'll do a more thorough look at this, along with the other math.

danieltjewett replied:
One thing, I would like to cache this data somewhere.  Once we calculate it once, it could help to not run this each time.

danieltjewett replied:
Though shoot, changing strings (like type writer effect, or other string changes) will have to recompute.  Hmm...

Hyreon replied:
I've added a cache. It seems to work great right now, since the only (provided) functions that change strings re-create a container every time they do. New container, new cache.
Ah, those containers may need to be deleted, but I don't think that's related to components. I'll leave that to you.

danieltjewett replied:
I wonder if instead of passing `useCache` (or whatever we rename it to be), we check to see if the original string is different than the previous string?  If it is different, we recompute.  More on this in my final notes, because there's something that this original Game Maker project isn't really doing that we are doing more in our internal game.

danieltjewett reviewed (changes requested):
It's getting there!  A few adjustments.  Then, I will dig into the math parts / double check within the internal game.

Hyreon committed:
Improved consistency

Improved consistency

danieltjewett commented on scripts/log/log.gml:
```diff
@@ -7,7 +7,7 @@ function log()
 	for (var i=0; i<argument_count; i++)
 	{
-	    str += string(argument[i]) + " ";
+	\tstr += string(argument[i]) + " ";
```
Something weird is going on here.

danieltjewett commented on scripts/update_data_with_components/update_data_with_components.gml:
```diff
@@ -0,0 +1,7 @@
+\tdata.components = get_components(data.str);
+\tdata.plaintext = string_join_ext("", array_map(data.components, function(component) {
+\t\treturn component.contents;
+\t}));
```
Something odd is going on here.

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+	
+	var last_width = 0;
+	
+	for (var i = 0; i < array_length(data.components) && !hit_character_limit;)
```
Let's move `array_length(data.components)` out of the loop (i.e. make a var called `componentsLength`.

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+		{
+			text_to_draw = carryover;
+			carryover = undefined;
+		}
```
Could we add a comment that explains this is prepending data from the previous iteration, if it exists?

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+						carryover = "";
+					}
+				}
+			}
```
Can you give me a scenario / example of how this logic runs / works?

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+				last_char = string_char_at(text_to_draw, string_length(text_to_draw));
+				carryover = last_char + carryover;
+				text_to_draw = string_delete(text_to_draw,string_length(text_to_draw),1);
+			}
```
I understand what is going on here, and I think it makes sense to check from the end to the beginning.  But man, `string_width` is slow, and `string_delete` probably isn't wonderful either.  I'm going to think on this, but there probably isn't a better way.

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+			}
+		}
+		
+		if (string_length(text_to_draw) > characters_remaining)
+		{
+			text_to_draw = string_copy(text_to_draw,0,characters_remaining);
+			characters_remaining = 0;
+		}
+		else
+		{
+			characters_remaining -= string_length(text_to_draw);
+		}
```
Let's add a comment and explain what is going on here.

danieltjewett replied:
Missing comment.

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+		}
+		if (characters_remaining <= 0) hit_character_limit = true;
+		
+		while (array_length(lines) <= line_index)
+		{
+			array_push(lines, {
+				width: 0,
+				segments: []
+			});
+		}
```
Why are we doing this? (Maybe a comment to help explain)

danieltjewett replied:
I think I understand now.  Each line can have multiple segments.  I think at the top of this file, making a comment of what the data structure of lines is would be helpful.  For example:
```
lines = [
  {
    width: number,
    segments: [
      {
        contents: string,
        emphasis: bool,
        remaining: string,
      },
      ...
    ]
  },
  ..
];
```

danieltjewett replied:
Could we add this to the top of the file?

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+	
+	var total_lines = array_length(lines);
+	
+	for (var line_index = 0; line_index < array_length(lines); line_index++)
```
Let's use `total_lines` instead of `array_length(lines)`.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+		var left_width = 0;
+		var right_width = line_width;
+		
+		for (var i = 0; i < array_length(lines[line_index].segments); i++)
```
let's make a var for `array_length(lines[line_index].segments)`.

danieltjewett replied:
Still needed.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+			if (text_to_draw != "")
+			{
+				
+				var size_of_segment = string_width(text_to_draw);
```
Given we did this a ton in `container_get_lines`, we ought to have this cached somewhere in the segment data structure.

danieltjewett replied:
This var also seems unused.

danieltjewett replied:
Is it possible to cache in `segement`?

danieltjewett replied:
Is it possible to cache in `segment`?

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+					width_offset = left_width; // fa_left or undefined
+				}
+				
+				text_to_draw = string_repeat("\n", line_index) + text_to_draw + string_repeat("\n", (total_lines - line_index));
```
Ha, clever!  Great way to not have to rework the entire `draw_container` system!  I wonder if we can store this in some kind of string accumulator instead, for efficiency purposes?

danieltjewett replied:
Still need to think about this.  I think this is something I can look at later.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+				right_width -= string_width(text_to_draw);
+				var width_offset;
+				
+				var halign = draw_get_halign();
```
We should be able to use `data.hAlign`.

danieltjewett commented on scripts/draw_container/draw_container.gml:
```diff
@@ -449,41 +449,95 @@ function draw_container(data)
+				
+				var halign = draw_get_halign();
+	
+				if (halign == fa_right)
+				{
+					width_offset = -right_width;
+				}
+				else if (halign == fa_center)
+				{
+					width_offset = (left_width-right_width)/2;
+				}
+				else
+				{
+					width_offset = left_width; // fa_left or undefined
+				}
```
I doubt we can do this, but I wonder if we can use the switch statement above (line ~402) to define `width_offset`?

Hyreon replied:
No, but we can use data.hAlign instead, which should be better than using `draw_get` calls

danieltjewett commented on scripts/draw_container_calculate/draw_container_calculate.gml:
```diff
@@ -2,6 +2,7 @@
 /// @param data
 function draw_container_calculate(data)
 {	
+	
```
Extra blank line.

danieltjewett commented on scripts/draw_container_get_properties/draw_container_get_properties.gml:
```diff
@@ -250,6 +255,11 @@ function draw_container_get_properties(data)
+	{
+		data.useCache = false; //whether to use cached lines
+	}
+	
```
I don't think `useCache` isn't very descriptive.  Maybe `recomputeString/LineData`?

danieltjewett commented on scripts/container_get_lines/container_get_lines.gml:
```diff
@@ -0,0 +1,133 @@
+	}
+	
+	data.segments = lines;
+	data.useCache = true;
```
This will override the setting passed into `draw_container`, thus, always caching.

Hyreon replied:
The original intention was that useCache would be overwritten by functions that change the data. However, with our new approach (that compares width and strings directly), changes to width or string will automatically re-cache the segments and components.

danieltjewett reviewed (changes requested):
Alright, loaded into the internal game.  Mostly good results.  However, a big issue because the `drawing container` project was built using the old concept of deleting the container every time, and reconstructing.  Let me explain.

Back in the day, the original way to "rerender" a container was just to `delete` and recreate.  This process would ensure a new container / rerender.  The problem is Game Maker is slow to construct these objects.  And, in most cases, nothing was changing from render to render.  Thus, we added the optimize example, which can be seen in `room_message10`.

The problem is that a lot of the logic assumes a brand new container each render.  Thus, an optimized container that changes its string, or does any other changes, won't recompute the lines / etc.  This is a major problem, but I think it's a pretty easy one to solve.

1. Check if the original string is different from the new string.  For things like the type writer effect, we would need to recompute `container_get_lines`, and any other associated strings going on (maybe the `get_components` function as well?).  I know I originally suggested the cache option, but I think this would be better and more practical.
2. I think we also need to recompute if `renderWidth` changes for a given container.  Much of the calculations are done with this variable, so we can also check if `previousRenderWidth` is != to `renderWidth` and if so, recompute.

There are probably other properties / things we need to double check.  A starting point would be to check what gets recomputed each time (`draw_container_calculate` as well as of course, `draw_container`).

The only other issue I saw was strings that contained `\n` were not rendering correctly with this added system.  i.e. I have a container in the form:

```
{
  hAnchor: fa_right,
  vAnchor: fa_top,
  grid: "column",
  children: [{
    grid: "column",
    requireFull: false,
    children: [
      {
	  font: font_DogicaPixel6Bold,
	  hAlign: fa_left,
	  marginBottom: INVENTORY_INFO_STAT_PADDING,
	  str: "Key",
      },
      {
	  font: font_DogicaPixel6,
	  hAlign: fa_left,
	  vAlign: fa_top,
	  marginBottom: INVENTORY_INFO_STAT_PADDING,
	  str: "Description balh adsf asfd j asdlfj.\n\nMore galsdf aasdf",
      }
    ],
    padding: INVENTORY_INFO_STAT_PADDING,
  }],
  x: global.GUI_WIDTH - INVENTORY_EDGE,
  y: INVENTORY_START_Y,
  sprite: spr_Inventory_Info_Box,
  imageAlpha: HELPER_DIALOG_ALPHA,
  width: sprite_get_width(spr_Inventory_Info_Box),
  height: sprite_get_height(spr_Inventory_Info_Box),
}
```

The string with `Description balh...` did not render correctly.

Hyreon committed:
Added requested improvements

Cache is now extended to components, not just segments; cache now updates whenever there is a change in str or width, and does not need to be reset by each function; removed instances of /t; added comments

danieltjewett reviewed (changes requested):
A couple of things from the previous review.  I also was trying to push code changes up, but it didn't work.  I'll try to either pair with you or screenshot the diffs.

Hyreon committed:
Updated caching

Cache is now saved at the end of every draw_container, and is more robust

Hyreon committed:
Strings now split by new lines correctly

danieltjewett commented:
Going to merge.  I have a few tweaks / refactors I can do myself.  Nice job with all of this!  This will help our game, and anyone using this project immensely!
